File: Sorting Algorithm/MergeSort.cs
// David Konopatskyi

using System;

class MergeSort
{
    static void Main(string[] args)
    {
        int[] numbers = new int[12] { 3, 6, -3, 10, 0, 12, 20, 66, 50, 5, 100, -10 };
        Console.WriteLine("\nOriginal Array Elements :");
        PrintArray(numbers);
        Console.WriteLine("\nSorted Array Elements :");
        PrintArray(Sort(numbers, 0, numbers.Length - 1));
        Console.WriteLine("\n");
    }

    static void Merge(int[] arr, int l, int m, int r)
    {
        // Find sizes of two subarrays to be merged 
        int n1 = m - l + 1;
        int n2 = r - m;

        /* Create temp arrays */
        int[] L = new int[n1];
        int[] R = new int[n2];

        /*Copy data to temp arrays*/
        for (int ii = 0; ii < n1; ++ii)
            L[ii] = arr[l + ii];
        for (int jj = 0; jj < n2; ++jj)
            R[jj] = arr[m + 1 + jj];

        /* Merge the temp arrays */

        // Initial indexes of first and second subarrays 
        int i = 0, j = 0;

        // Initial index of merged subarry array 
        int k = l;
        while (i < n1 && j < n2)
        {
            if (L[i] <= R[j])
            {
                arr[k] = L[i];
                i++;
            }
            else
            {
                arr[k] = R[j];
                j++;
            }
            k++;
        }

        /* Copy remaining elements of L[] if any */
        while (i < n1)
        {
            arr[k] = L[i];
            i++;
            k++;
        }

        /* Copy remaining elements of R[] if any */
        while (j < n2)
        {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    static int[] Sort(int[] arr, int l, int r)
    {
        if (l < r)
        {
            // Find the middle point 
            int m = (l + r) / 2;

            // Sort first and second halves 
            Sort(arr, l, m);
            Sort(arr, m + 1, r);

            // Merge the sorted halves 
            Merge(arr, l, m, r);
        }
        return arr;
    }

    static void PrintArray(int[] array)
    {
        foreach (int i in array)
        {
            Console.Write(i.ToString() + "  ");
        }
    }

}

File: Sorting Algorithm/Java/MergeSort.java
//Code for Merge Sort for type  int in Java
//Author: @BoaMos

import java.io.*;

public class MergeSort { 
	// Merges two arrays of arr[] 
	//Merge such that L  Element  R
   public void merge(int arr[], int l, int m, int r) 
   { 
   	// Find index of L an R
      int LIndex = m - l + 1; 
      int RIndex = r - m; 
   	
      // Create temp arrays 
      int L[] = new int[LIndex]; 
      int R[] = new int[RIndex]; 
   
   	//add data to arrays
      for (int i = 0; i < LIndex; ++i) 
         L[i] = arr[l + i];
          
      for (int j = 0; j < RIndex; ++j) 
         R[j] = arr[m + 1 + j]; 
   
   	//Merge  arrays
   	// Initial indexes of first and second subarrays 
      int i = 0, j = 0; 
   
   	// Initial index of merged subarry array 
      int k = l; 
      while (i < LIndex && j < RIndex) { 
         if (L[i] <= R[j]) { 
            arr[k] = L[i]; 
            i++; 
         } 
         else { 
            arr[k] = R[j]; 
            j++; 
         } 
         k++; 
      } 
   
   	// Put l[] in sequence 
      while (i < LIndex) { 
         arr[k] = L[i]; 
         i++; 
         k++; 
      } 
   
   	// copy R[] to sequence
      while (j < RIndex) { 
         arr[k] = R[j]; 
         j++; 
         k++; 
      } 
   } 

	/*
   *para ([] arr)>> sorted array
   *para (int l) >> index of left which is < mid 
   *para (int r)>> index of right which is > mid 
   */
   public void sort(int arr[], int l, int r) 
   { 
      if (l < r) { 
      	// Find the middle point 
         int mid = (l + r) / 2; 
      
      	// Sort left 
         sort(arr, l, mid);
         // Sort Right  
         sort(arr, mid + 1, r); 
      
      	// Merge the sorted halves 
         merge(arr, l, mid, r); 
      } 
   } 

	/*
   *main method 
   */ 
   public static void main(String args[]) throws Exception{ 
      int arr[] = { 82,5,85,65,848,8,62,41,2,0 }; 
   
   //print out the array 
      System.out.print("Unsorted array>> "); 
      for (int i =0; i <= arr.length-1; i++)
         System.out.print(arr[i]+" "); 
   
      MergeSort tool = new MergeSort(); 
      tool.sort(arr, 0, arr.length - 1); 
   
      System.out.print("\nSorted array >> "); 
      for (int i =0; i <= arr.length-1; i++)
         System.out.print(arr[i]+" ");   
   
   
   //write to file 
      FileWriter fl= new FileWriter (new File ("Array.txt"));
      PrintWriter pr = new PrintWriter (fl);
   
      for (int e =0 ; e<= arr.length ; e++)
         pr.println(arr[e]);
   
      pr.close();
   
   }
   
   
} 

